### ISSUE TITLE
Regarding the Terminal Option in the Compose-UI

### ISSUE THREAD

**Issue by mkscala** [francescou/docker-compose-ui#4]
Hi,
 Is is possible to add a terminal(tty) embedded in the UI which will help us to checks the logs and run commands within the browser.

Thanks,
Mani

**Comment by francescou** [comment]
Hello Mani,

it could be interesting to integrate a TTY in the application. I will investigate about the feasibility of this feature ASAP

**Comment by francescou** [comment]
there is no standard way to open a terminal using docker-py (see https://github.com/docker/docker-py/issues/390, https://github.com/docker/docker-py/pull/239, https://github.com/docker/docker-py/issues/389).
Maybe it could be possible using https://github.com/d11wtq/dockerpty (it is already used by docker-compose: https://github.com/docker/compose/blob/master/requirements.txt#L4) but I didn't find a way to perform a `docker exec -it $CONTAINER /bin/bash` on a running container (there is also an issue about this: https://github.com/d11wtq/dockerpty/issues/30)

**Comment by WTFKr0** [comment]
Hi !

Any news on that feature ?

**Comment by francescou** [comment]
I did some test with the latest docker pty
 (see commit 258b7e0d0be7c91df874f96e41b4af0e90dbddc1)

There is no GUI at the moment but you can try it from an HTTP client:

```
curl -X POST http://localhost:5000/api/v1/terminal -H'Content-type: application/json' --data '{"command":"ls -lrt /", "name":"node-redis", "id":"noderedis_redis_1"}'
```

**Comment by francescou** [comment]
added experimental UI in 4b040de2a18c0328deba203566f76fe385a74f06

cc @dvdred

**Comment by francescou** [comment]
current solution, based on `exec_create` and `exec_start` is too limiting: there is no way to run interactive commands (e.g. redis-cli). I guess a better alternative would be to use websockets.

**Comment by allamand** [comment]
Hi Guys,

You can perform this using shipyard project you can set-up on top of existing docker or swarm cluster.

I used this compose to launch shipyard with docker-ui on a swarm with a Traefik proxy for routing

```
version: '2'

services:
 rethinkdb:
    image: rethinkdb:latest
    labels:
      - "traefik.enable=false"
    networks:
      - back

 shipyard:
    image: shipyard/shipyard:latest
    container_name: shipyard-controller
    restart: always
    expose:
      - "8080"
    depends_on: ["rethinkdb"]
    volumes:
      - /etc/docker/swarm:/ssl:ro
    command: |
      --debug
      server 
      -d tcp://10.234.172.115:3376 
      --tls-ca-cert=/ssl/ca.pem 
      --tls-cert=/ssl/server.pem 
      --tls-key=/ssl/server-key.pem
    labels:
      - "traefik.frontend.rule=Host:shipyard.mydomain.com"
      - "traefik.docker.network=traefik_net"
    networks:
      - traefik_net
      - back

networks:
  traefik_net:
      external: true
  back:
    driver: overlay
```

**Comment by mkscala** [comment]
@allamand , shipyard integration gives an CLI or  Unix like web Terminal UI ?

**Comment by allamand** [comment]
@murugesh it's a unix like terminal ui (a console in the browser) I can show example if you like, I launch my shipyard using compose-ui

Sébastien

**Comment by mkscala** [comment]
I would live to see the example, how can you show me?

**Comment by mkscala** [comment]
I would love to see the example, how can you show me?

**Comment by allamand** [comment]
I used this compose file : https://gist.github.com/allamand/e4f6a6229804007fa8f1a37761745031.

I din't saw the entire video but it should create something like this https://www.youtube.com/watch?v=aLlp-kemx_s

**Comment by MichaelMackus** [comment]
Looks like there's an API method in Docker Engine which exposes the raw stream from the docker exec. [Here](https://github.com/shipyard/shipyard/blob/master/controller/api/exec.go) is the relevant code from Shipyard which seems to just connect to that endpoint & open it up with websockets.

@francescou would you accept PRs for similar functionality? Seems like its all implemented in Docker Engine, so the missing piece would be the Websocket connection (likely from some Python websocket library).

**Comment by francescou** [comment]
@MichaelMackus I would be really interested. I've also made some experiments with websockets in python/Flask (using gevent, see #14) but I didn't like how these framework worked together.

**Comment by MichaelMackus** [comment]
Good to hear! I did try some experimenting with Websockets as well, but I'm still trying to wrap my head around the protocol (& integrating it with Flask). I'll let you know what I find.

**Comment by francescou** [comment]
This feature has been released in `francescou/docker-compose-ui:1.8.0` thanks to #91 

instructions:
- download [docker-compose.yml](https://github.com/francescou/docker-compose-ui/blob/v1.8.0/docker-compose.yml)
- run `docker-compose up`
- open your browser to <http://localhost:8080>
